ms2rescore/feature_generators/ms2.py

```diff
@@ -0,0 +1,319 @@
+            "matched_y_ions",
+            "matched_y_ions_pct",
+            "matched_ions_pct",
+            "hyperscore",
+        ]
+
+    def add_features(self, psm_list: PSMList) -> None:
+        """Add MS2-derived features to PSMs."""
+
+        logger.info("Adding MS2-derived features to PSMs.")
+        psm_dict = psm_list.get_psm_dict()
+        current_run = 1
+        total_runs = sum(len(runs) for runs in psm_dict.values())
+
+        for runs in psm_dict.values():
+            for run, psms in runs.items():
+                logger.info(
+                    f"Running MS2 for PSMs from run ({current_run}/{total_runs}) `{run}`..."
+                )
+                psm_list_run = PSMList(psm_list=list(chain.from_iterable(psms.values())))
+                spectrum_filename = infer_spectrum_path(self.spectrum_path, run)
+
+                self._calculate_features(psm_list_run, spectrum_filename)
+                current_run += 1
+
+    def _calculate_features(self, psm_list: PSMList, spectrum_file: str) -> List:
+        """Retrieve annotated spectra for all psms."""
+
+        spectrum_reader = self._read_spectrum_file(spectrum_file)
+
+        spectrum_id_pattern = re.compile(
+            self.spectrum_id_pattern if self.spectrum_id_pattern else r"(.*)"
+        )
+
+        try:
+            mapper = {
+                spectrum_id_pattern.search(spectrum_id).group(1): spectrum_id
+                for spectrum_id in spectrum_reader._offset_index.mapping["spectrum"].keys()
+            }
+        except AttributeError:
+            raise ParseSpectrumError(
+                "Could not parse spectrum IDs using ´spectrum_id_pattern´. Please make sure that there is a capturing in the pattern."
+            )
+        annotated_spectra = [
+            self._annotate_spectrum(psm, spectrum_reader.get_by_id(mapper[psm.spectrum_id]))
+            for psm in psm_list
+        ]
+        for psm, annotated_spectrum in zip(psm_list, annotated_spectra):
+            psm.rescoring_features.update(
+                self._calculate_spectrum_features(psm, annotated_spectrum)
+            )
+
+            if self.calculate_hyperscore:
+                # Filters out peaks which are unnannotated (can be specified, but keep at b-y ions of any charge ?)
+                b, y = get_by_fragments(annotated_spectrum)
+                hs = calculate_hyperscore(
+                    n_y=len(y),
+                    n_b=len(b),
+                    y_ion_intensities=y,
+                    b_ion_intensities=b
+                )
+                psm.rescoring_features.update({'hyperscore': hs})
+
+    @staticmethod
+    def _read_spectrum_file(spectrum_filepath: Path) -> Union[mzml.PreIndexedMzML, mgf.IndexedMGF]:
+
+        if spectrum_filepath.suffix.lower() == ".mzml":
+            return mzml.PreIndexedMzML(str(spectrum_filepath))
+        elif spectrum_filepath.suffix.lower() == ".mgf":
+            return mgf.IndexedMGF(str(spectrum_filepath))
```

Comment:
It might be better to avoid failing silently and add an else and raise an e.g. `NotImplementedError` or `ValueError`.